opm/simulators/linalg/bda/rocsparseSolverBackend.cpp

```diff
@@ -0,0 +1,626 @@
+template <unsigned int block_size>
+bool rocsparseSolverBackend<block_size>::create_preconditioner() {
+    Timer t;
+
+    bool result = true;
+    ROCSPARSE_CHECK(rocsparse_dbsrilu0(handle, dir, Nb, nnzbs_prec, descr_M,
+                    d_Mvals, d_Mrows, d_Mcols, block_size, ilu_info, rocsparse_solve_policy_auto, d_buffer));
+
+    // Check for zero pivot
+    int zero_position = 0;
+    rocsparse_status status = rocsparse_bsrilu0_zero_pivot(handle, ilu_info, &zero_position);
+    if(rocsparse_status_success != status)
+    {
+        printf("L has structural and/or numerical zero at L(%d,%d)\n", zero_position, zero_position);
+        return false;
+    }
+
+    if (verbosity >= 3) {
+        HIP_CHECK(hipStreamSynchronize(stream));
+        std::ostringstream out;
+        out << "rocsparseSolver::create_preconditioner(): " << t.stop() << " s";
+        OpmLog::info(out.str());
+    }
+    return result;
+} // end create_preconditioner()
+
+
+template <unsigned int block_size>
+void rocsparseSolverBackend<block_size>::solve_system(WellContributions &wellContribs, BdaResult &res) {
+    Timer t;
+
+    // actually solve
+    try {
+        gpu_pbicgstab(wellContribs, res);
+    } catch (const cl::Error& error) {
+        std::ostringstream oss;
+        oss << "rocsparseSolverBackend::solve_system error: " << error.what() << "(" << error.err() << ")\n";
+        oss << getErrorString(error.err());
+        // rethrow exception
+        OPM_THROW(std::logic_error, oss.str());
+    } catch (const std::logic_error& error) {
+        // rethrow exception by OPM_THROW in the try{}, without this, a segfault occurs
+        throw error;
+    }
+
+    if (verbosity >= 3) {
+        HIP_CHECK(hipStreamSynchronize(stream));
+        std::ostringstream out;
+        out << "rocsparseSolver::solve_system(): " << t.stop() << " s";
+        OpmLog::info(out.str());
+    }
+
+} // end solve_system()
+
+
+// copy result to host memory
+// caller must be sure that x is a valid array
+template <unsigned int block_size>
+void rocsparseSolverBackend<block_size>::get_result(double *x) {
+    Timer t;
+
+    HIP_CHECK(hipMemcpyAsync(x, d_x, sizeof(double) * N, hipMemcpyDeviceToHost, stream));
+    HIP_CHECK(hipStreamSynchronize(stream)); // always wait, caller might want to use x immediately
+
+    if (verbosity >= 3) {
+        std::ostringstream out;
+        out << "rocsparseSolver::get_result(): " << t.stop() << " s";
+        OpmLog::info(out.str());
+    }
+} // end get_result()
+
+
+template <unsigned int block_size>
+SolverStatus rocsparseSolverBackend<block_size>::solve_system(std::shared_ptr<BlockedMatrix> matrix,
+                                                           double *b,
+                                                           std::shared_ptr<BlockedMatrix> jacMatrix,
+                                                           WellContributions& wellContribs,
+                                                           BdaResult &res)
```

Comment:
Sorry for nitpicking: Indentation seems a bit off.

Reply:
Fixed in a squash